vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp

```diff
@@ -177,12 +180,13 @@ void ReactInstanceWin::Initialize() noexcept {
         // Objects that must be created on the UI thread
         if (auto strongThis = weakThis.GetStrongPtr()) {
           auto const &legacyInstance = strongThis->m_legacyReactInstance;
-          strongThis->m_deviceInfo = std::make_shared<react::uwp::DeviceInfo>(legacyInstance);
           strongThis->m_appTheme =
               std::make_shared<react::uwp::AppTheme>(legacyInstance, strongThis->m_uiMessageThread.LoadWithLock());
           react::uwp::I18nHelper().Instance().setInfo(react::uwp::I18nModule::GetI18nInfo());
           strongThis->m_appearanceListener =
               Mso::Make<react::uwp::AppearanceChangeListener>(legacyInstance, strongThis->m_uiQueue);
+          ::Microsoft::ReactNative::DeviceInfoHolder::InitDeviceInfoHolder(
```

Comment:
>InitDeviceInfoHolder [](start = 54, length = 20)

It would be nice to discuss how we want to do it in future. Ideally we should do the extra UI-bound work only for modules that are passed to the instance and it would be great that the ReactInstance does not know them 'by name'.
I think we can add a couple of things:
- Each module is allowed to claim that it wants to initialize some data on UI thread before it is instantiated.
- We do all the initialization before we initialize the ReactNativeHost to allow it to do all the work in a background thread.

Reply:
Another alternative is to do all this work in a module provider and require all providers to be created in UI thread.

---
In reply to: [429379351](https://github.com/microsoft/react-native-windows/pull/4991#discussion_r429379351) [](ancestors = 429379351)